### PR TITLE
EC/CUDA: vectorize reduce kernels

### PR DESCRIPTION
**Pull request by samnordmann**
## What
Optimize reduce kernels by using vectorial loads. It allows substantial improvement in the band-width, especially for small datatypes.
Data for triggered TL/UCP Allreduce on CUDA buffers between two processes on the same node
allreduce_triggered=True_plot
The optimization is only enabled for triggered operations. For non-triggered operations, there is no change in the perfs:
allreduce_triggered=False_plot

### CODE REVIEW

**Comment by samnordmann** [comment]
Further perf measurements:
reducedt_master_nt=256nt=256_plot
reducedt_master_nt=256-T-Tnt=256_plot
reducedt_master_nt=512nt=512_plot
reducedt_master_nt=512-T-Tnt=512_plot
reducedt_strided_master_nt=256nt=256_plot
reducedt_strided_master_nt=256-T-Tnt=256_plot
reducedt_strided_master_nt=512nt=512_plot
reducedt_strided_master_nt=512-T-Tnt=512_plot

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h [hunk 1]
```diff
@@ -59,103 +61,152 @@ cuFloatComplex operator* (const cuFloatComplex & first,
...
             memcpy(s, task_default_p->srcs,                                     \
-                   UCC_EE_EXECUTOR_NUM_BUFS * sizeof(_Type *));                 \
+                   UCC_EE_EXECUTOR_NUM_BUFS * sizeof(Type *));                  \
             n_src2 = task_default_p->n_srcs - 1;                                \
             s1     = s[0];                                                      \
+            for (int i = 0; i < MAXSRCS && i <= n_src2; i++)                    \
+                alignedVec |= ptrAlignVec<Type, vectype>(s[i]);                 \
+        }                                                                       \
+        alignedVec |= ptrAlignVec<Type, vectype>(d);                            \
+        alignedVec |= sizeof(vectype) % sizeof(Type) != 0;                      \
```
[the remark this answers is not in the document]
Comment (samnordmann) [reply src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h:186]: ok, I replaced it with `ucc_assert_system`

File: src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h [hunk 2]
```diff
@@ -59,103 +61,152 @@ cuFloatComplex operator* (const cuFloatComplex & first,
...
+                      : (blockDim.x * gridDim.x) / unroll_group_size;          \
+        const int    idx = threadIdx.x % unroll_group_size;                    \
+        const size_t start =                                                   \
+            offset + (group * unroll_group_size * unroll + idx) * vectorize;   \
+        const size_t step =                                                    \
+            num_groups * unroll_group_size * unroll * vectorize;               \
+        const size_t end =                                                     \
+            offset +                                                           \
+            ((count - offset) / (unroll_group_size * unroll * vectorize)) *    \
+                (unroll_group_size * unroll * vectorize);                      \
```
[the remark this answers is not in the document]
Comment (samnordmann) [reply src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h:93]: done, defining a variable `const int factor`

File: src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h [hunk 3]
```diff
@@ -59,103 +61,152 @@ cuFloatComplex operator* (const cuFloatComplex & first,
...
+            /* if buffers align, use vectorized loads and unrolling */          \
+            CUDA_REDUCE_WITH_OP_CHUNK(UNROLL, WARP_SIZE, _OP, vectype);         \
+            /* call with vectorization but without unrolling */                 \
+            CUDA_REDUCE_WITH_OP_CHUNK(1, 1, _OP, vectype);                      \
         }                                                                       \
-        CUDA_REDUCE_WITH_OP_CHUNK(0, UNROLL, WARP_SIZE, _OP);                   \
-        /* second call for data remainder */                                    \
-        CUDA_REDUCE_WITH_OP_CHUNK(                                              \
-            (count / (WARP_SIZE * UNROLL)) * (WARP_SIZE * UNROLL), 1, 1, _OP);  \
+        /* call with unrolling but without vectorization */                     \
```
[the remark this answers is not in the document]
Comment (samnordmann) [reply src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h:193]: comment added

**Comment by manjugv** [comment]
ping @bureddy

**Review by bureddy — APPROVED** [review]

**Pull request merged**